### PR TITLE
refactor: move the rest of block pre-processing logic to chain

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -3008,17 +3010,18 @@ impl Chain {
...
                 &prev_block,
                 &receipts_by_shard,
                 ApplyChunksMode::CatchingUp,
+                None,
             )?;
             blocks_catch_up_state
                 .scheduled_blocks
-                .insert(pending_block, chain_update.into_saved_store_update());
+                // TODO: no need for a saved store update here at all? o_O
+                .insert(pending_block, self.chain_update().into_saved_store_update());
```
[the remark this answers is not in the document]
Comment (mzhangmzz) [reply chain/chain/src/chain.rs:3024]: I think you are right here! That's good. More code simplification.

File: chain/chain/src/chain.rs [hunk 2]
```diff
@@ -3181,147 +3176,545 @@ impl Chain {
...
-            (Some(h1), Some(h2)) => Some(combine_hash(&h1, &h2)),
-            (Some(h1), None) => Some(h1),
-            (None, Some(_)) => {
-                debug_assert!(false, "Inconsistent state in merkle proof computation: left node is None but right node exists");
-                None
-            }
-            _ => None,
-        }
-    }
+    pub fn create_chunk_state_challenge(
```
Comment (mm-near) [inline chain/chain/src/chain.rs:3180]: due to code moves -- the git shows this file as 2k diff - very had to see which parts of the functions where 'just' moved vs modified.

**Review by mm-near — CHANGES_REQUESTED** [review]

File: core/primitives/src/state_patch.rs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+use crate::state_record::StateRecord;
+
+pub struct StatePatch {
```
[the remark this answers is not in the document]
Comment (matklad) [reply core/primitives/src/state_patch.rs:3]: wait, looks like I've messed up my git branching, this shouldn't be here yet at all..

[the remark this answers is not in the document]
Comment (matklad) [reply core/primitives/src/state_patch.rs]: Ah no, that's not a messed up branched, just accidentally commited code. The better version would be in a separate PR: https://github.com/near/nearcore/pull/6967

**Comment by matklad** [comment]
```console
$ git diff --color-moved=dimmed-zebra HEAD~11 HEAD
```

Is the best I can come up with to review the code here :(

**Review by mm-near — APPROVED** [review]

**Pull request merged**